File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/logic/ServiceCalendar.java
package com.example.prj_gifu_univ_bus_navi.logic;

import com.example.prj_gifu_univ_bus_navi.data.LocalAcademicCalendarData;
import com.example.prj_gifu_univ_bus_navi.data.LocalHolidayData;
import com.example.prj_gifu_univ_bus_navi.model.BaseDayType;
import com.example.prj_gifu_univ_bus_navi.model.ServiceDateContext;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.List;

public final class ServiceCalendar {
    private ServiceCalendar() {
    }

    public static ServiceDateContext createContext(LocalDate date, List<LocalDate> schoolHolidays) {
        BaseDayType baseDayType;
        if (LocalHolidayData.isHoliday(date)
            || date.getDayOfWeek() == DayOfWeek.SATURDAY
            || date.getDayOfWeek() == DayOfWeek.SUNDAY) {
            baseDayType = BaseDayType.WEEKEND_HOLIDAY;
        } else {
            baseDayType = BaseDayType.WEEKDAY;
        }
        return new ServiceDateContext(
            date,
            baseDayType,
            LocalAcademicCalendarData.isSchoolHoliday(date),
            date.getMonthValue()
        );
    }
}

File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/logic/ServiceCalendar.kt
package com.example.prj_gifu_univ_bus_navi.logic

import com.example.prj_gifu_univ_bus_navi.model.BaseDayType
import com.example.prj_gifu_univ_bus_navi.model.ServiceDateContext
import java.time.DayOfWeek
import java.time.LocalDate

object ServiceCalendar {
    fun createContext(date: LocalDate, schoolHolidays: List<LocalDate>): ServiceDateContext {
        val baseDayType = when (date.dayOfWeek) {
            DayOfWeek.SATURDAY, DayOfWeek.SUNDAY -> BaseDayType.WEEKEND_HOLIDAY
            else -> BaseDayType.WEEKDAY
        }
        return ServiceDateContext(
            date = date,
            baseDayType = baseDayType,
            isSchoolHoliday = date in schoolHolidays,
            month = date.monthValue,
        )
    }
}
